onnxruntime/contrib_ops/cpu/optional_ops/optional_ops.cc

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "optional_ops.h"
+#include "core/framework/ml_value.h"
+#include "core/providers/cpu/tensor/utils.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(Optional, kMSDomain, 1, kCpuExecutionProvider,
+                        KernelDefBuilder()
+                            .TypeConstraint("V", DataTypeImpl::AllTensorAndSequenceTensorTypes())
+                            .TypeConstraint("O", DataTypeImpl::AllOptionalTypes())
+                            // We may be able to re-use the input for the output as is unless the output
+                            // is a graph output. We provide this hint to the allocation planner
+                            // to make the re-use call.
+                            .Alias(0, 0),
+                        Optional);
+
+ONNX_OPERATOR_KERNEL_EX(OptionalHasElement, kMSDomain, 1, kCpuExecutionProvider,
+                        KernelDefBuilder()
+                            .TypeConstraint("O", DataTypeImpl::AllOptionalTypes())
+                            .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>()),
+                        OptionalHasElement);
+
+ONNX_OPERATOR_KERNEL_EX(OptionalGetElement, kMSDomain, 1, kCpuExecutionProvider,
+                        KernelDefBuilder()
+                            .TypeConstraint("O", DataTypeImpl::AllOptionalTypes())
+                            .TypeConstraint("V", DataTypeImpl::AllTensorAndSequenceTensorTypes())
+                            // We may be able to re-use the input for the output as is unless the output
+                            // is a graph output. We provide this hint to the allocation planner
+                            // to make the re-use call.
+                            .Alias(0, 0),
+                        OptionalGetElement);
+
+static void CloneTensor(AllocatorPtr alloc, const Tensor& tensor_to_be_cloned, OrtValue& clone) {
+  auto tensor = std::make_unique<Tensor>(tensor_to_be_cloned.DataType(),
+                                         TensorShape(tensor_to_be_cloned.Shape()), alloc);
+  CopyCpuTensor(&tensor_to_be_cloned, tensor.get());
```

Comment:
All other node output allocations are handled within ExecutionFrame and take the allocation plan into consideration. You've pulled this allocation out to the operator implementation where the allocation plan isn't available, and are using the temp space allocator for it. Is that safe, or should we instead add a new 'Output' method to OpKernelContext that supports this usage whilst keeping the allocation within ExecutionFrame?

Reply:
Indirectly, this node allocation is controlled by the allocation plan as well. We expect that in almost all cases the input OrtValue ill be re-used as outputs (.Alias(0,0)) for `Optional` and `OptionalGetElement`. There are comments along the way in these ops' implementation to explain this. In the rare case that for some reason, we are unable to re-use the input OrtValue as the output (AFAIK this will happen only when the outputs of these nodes are graph outputs), only then will we invoke the `Clone...()` methods based on the actual type housed in the input OrtValue.

As for using the temp space allocator, that is how we handle allocations for tensors within sequence tensors even today. For tensors - I am open to suggestions if there is a better way to do the buffer allocation, as far as I can tell, even when the execution frame does an allocation for a memory owned tensor (when the allocation plan does not point to a memory buffer re-use), it uses the allocator from the session state. Should, we do something similar here ?

Reply:
It's the 'indirectly' part that concerns me as it's changing the setup from the ExecutionFrame managing all allocations, to it managing most allocations. That means a developer checking allocation logic now has more things/places they need to be aware of and reason about. 

If there were methods provided by OpKernelContext to check if an input has data and to create an empty optional output could we simply use the normal Input/Output API?

Reply:
Fixed